src/algokit/core/dispenser.py

```diff
@@ -0,0 +1,199 @@
+KEYRING_KEY = "algokit_account"
+DISPENSER_REQUEST_TIMEOUT = 15
+
+
+class DispenserApiAudiences(str, Enum):
+    USER = "staging-dispenser-api-user "
+    CI = "staging-dispenser-api-ci"
+
+
+@dataclass
+class AuthConfig:
+    domain: str
+    audiences: dict[DispenserApiAudiences, str]
+    client_ids: dict[DispenserApiAudiences, str]
+    base_url: str
+
+
+@dataclass
+class AccountKeyringData:
+    id_token: str
+    access_token: str
+    refresh_token: str
+    user_id: str
+
+
+AUTH_CONFIG = AuthConfig(
+    domain="dispenser-staging.eu.auth0.com",
+    audiences={
+        DispenserApiAudiences.USER: "api-staging-dispenser-user",
+        DispenserApiAudiences.CI: "api-staging-dispenser-ci",
+    },
+    client_ids={
+        DispenserApiAudiences.USER: "flwPVx0HstfeZtGd3ZJVSwhGCFlR5v8a",
+        DispenserApiAudiences.CI: "q3EWUsOmj5jINIchDxnm9gMEa10Th7Zj",
+    },
+    base_url="https://2i2vncra6a.execute-api.us-east-1.amazonaws.com",
+)
+
+
+def process_dispenser_request(
+    *, url_suffix: str, data: dict | None = None, ci: bool, method: str = "POST"
+) -> httpx.Response | None:
+    """Generalized method to process http requests to dispenser API"""
+
+    if not ci and not is_authenticated():
+        logger.error("Please login first")
+        return None
+
+    headers = {"Authorization": f"Bearer {get_auth_token(ci=ci)}"}
+
+    # Set request arguments
+    request_args = {
+        "url": f"{AUTH_CONFIG.base_url}/{url_suffix}",
+        "headers": headers,
+        "timeout": DISPENSER_REQUEST_TIMEOUT,
+    }
+
+    if method.upper() != "GET" and data is not None:
+        request_args["json"] = data
+
+    try:
+        response: httpx.Response = getattr(httpx, method.lower())(**request_args)
+        return response
+    except Exception as err:
+        error_message = "Error processing dispenser API request"
+        logger.debug(f"{error_message}: {err}", exc_info=True)
+        raise Exception(error_message) from err
+
+
+def set_keyring_passwords(token_data: dict[str, str]) -> None:
+    decoded_id_token = jwt.decode(token_data["id_token"], algorithms=ALGORITHMS, options={"verify_signature": False})
+    user_id = decoded_id_token.get("sub")
+    data = AccountKeyringData(
+        id_token=token_data["id_token"],
+        access_token=token_data["access_token"],
+        refresh_token=token_data.get("refresh_token", ""),
+        user_id=user_id or "",
+    )
+    keyring.set_password(KEYRING_NAMESPACE, KEYRING_KEY, json.dumps(data.__dict__))
+
+
+def get_keyring_data() -> AccountKeyringData:
+    data_str = keyring.get_password(KEYRING_NAMESPACE, KEYRING_KEY)
+    if not data_str:
+        raise Exception("No keyring data found")
+    return AccountKeyringData(**json.loads(data_str))
+
+
+def is_authenticated() -> bool:
+    try:
+        data = get_keyring_data()
+    except Exception:
+        return False
+    decoded_token = jwt.decode(data.access_token, options={"verify_signature": False})
```

Comment:
What's the reason for not verifying the signature?

Reply:
@neilcampbell so apparently auth0 client in python only concerned about an interface around management api which doesn't include validation of access tokens. Had to add that manually by assembling rsa pub key and running validate. check is_authenticated for latest on how verify_signature is now set to True